test/tactic_set.lean

```diff
@@ -0,0 +1,389 @@
+by set_taut
+
+theorem union_diff_distrib {s t u : set α} : (s ∪ t) \ u = s \ u ∪ t \ u :=
+by set_taut'
+
+theorem inter_union_distrib_left {s t u : set α} : s ∩ (t ∪ u) = (s ∩ t) ∪ (s ∩ u) := by set_taut'
+
+theorem inter_union_distrib_right {s t u : set α} : (s ∩ t) ∪ u = (s ∪ u) ∩ (t ∪ u) := by set_taut'
+
+theorem union_inter_distrib_left {s t u : set α} : s ∪ (t ∩ u) = (s ∪ t) ∩ (s ∪ u) :=  by set_taut'
+
+theorem union_inter_distrib_right {s t u : set α} : (s ∪ t) ∩ u = (s ∩ u) ∪ (t ∩ u) := by set_taut'
+
+theorem inter_diff_assoc (a b c : set α) : (a ∩ b) \ c = a ∩ (b \ c) :=
+by set_taut
+
+theorem inter_diff_self (a b : set α) : a ∩ (b \ a) = ∅ :=
+by set_taut
+
+theorem inter_union_diff (s t : set α) : (s ∩ t) ∪ (s \ t) = s := by set_taut
+
+theorem inter_union_compl (s t : set α) : (s ∩ t) ∪ (s ∩ tᶜ) = s := by set_taut
+
+
+
+theorem compl_eq_univ_diff (s : set α) : sᶜ = univ \ s :=
+by set_taut
+
+lemma empty_diff (s : set α) : (∅ \ s : set α) = ∅ := by set_taut
+
+
+theorem diff_empty {s : set α} : s \ ∅ = s := by set_taut
+
+theorem diff_diff {u : set α} : s \ t \ u = s \ (t ∪ u) := by set_taut
+
+-- the following statement contains parentheses to help the reader
+lemma diff_diff_comm {s t u : set α} : (s \ t) \ u = (s \ u) \ t := by set_taut
+
+lemma diff_inter {s t u : set α} : s \ (t ∩ u) = (s \ t) ∪ (s \ u) :=
+by set_taut'
+
+lemma diff_inter_diff {s t u : set α} : s \ t ∩ (s \ u) = s \ (t ∪ u) :=
+by set_taut'
+
+lemma diff_compl : s \ tᶜ = s ∩ t := by set_taut'
+
+lemma diff_diff_right {s t u : set α} : s \ (t \ u) = (s \ t) ∪ (s ∩ u) :=
+by set_taut'
+
+theorem insert_diff_of_mem (s) (h : a ∈ t) : insert a s \ t = s \ t :=
+by by set_taut'
+
+lemma insert_diff_self_of_not_mem {a : α} {s : set α} (h : a ∉ s) :
+  insert a s \ {a} = s :=
+by set_taut'
+
+
+theorem union_diff_self {s t : set α} : s ∪ (t \ s) = s ∪ t :=
+by set_taut'
+
+theorem diff_union_self {s t : set α} : (s \ t) ∪ t = s ∪ t :=
+by set_taut'
+
+theorem diff_inter_self {a b : set α} : (b \ a) ∩ a = ∅ :=
+by set_taut'
+
+theorem diff_inter_self_eq_diff {s t : set α} : s \ (t ∩ s) = s \ t :=
+by set_taut'
+
+theorem diff_self_inter {s t : set α} : s \ (s ∩ t) = s \ t :=
+by set_taut'
+
+theorem diff_singleton_eq_self {a : α} {s : set α} (h : a ∉ s) : s \ {a} = s := by set_taut'
+
+theorem insert_diff_singleton {a : α} {s : set α} :
+  insert a (s \ {a}) = insert a s := by set_taut'
+
+lemma diff_self {s : set α} : s \ s = ∅ := by set_taut'
+
+lemma diff_diff_cancel_left {s t : set α} (h : s ⊆ t) : t \ (t \ s) = s :=
+by set_taut'
+
+end set_tests
+
+namespace set_tests
+variables {α β γ:Type*}
+section preimage
+variables {f : α → β} {g : β → γ}
+
+theorem preimage_empty : f ⁻¹' ∅ = ∅ := by set_taut
+
+
+lemma preimage_congr {f g : α → β} {s : set β} (h : ∀ (x : α), f x = g x) : f ⁻¹' s = g ⁻¹' s := by set_taut
+
+theorem preimage_univ : f ⁻¹' univ = univ := by set_taut
+
+theorem subset_preimage_univ {s : set α} : s ⊆ f ⁻¹' univ := subset_univ _
```

Comment:
This proof doesn't use the `set_taut` tactic, is it included in the test file for some reason? Possibly you would like it to be handled by a future version of the tactic? In that case, you should mark it with a TODO comment stating this.

Reply:
I was trying to figure out how to automatically prove something is a subset. I am a big noob when it comes to the metaprogramming language, and couldn't figure out the strategy.